### PR TITLE
Rewrite request hooks to TypeScript

### PR DESCRIPTION
**Pull request by miherlosev**

### CODE REVIEW

**Comment by miherlosev** [comment]
@testcafe-build-bot retest

**Comment by miherlosev** [comment]
@testcafe-build-bot retest

**Comment by miherlosev** [comment]
FPR

File: src/api/request-hooks/hook.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
+import {
+    RequestFilterRule,
+    ConfigureResponseEvent,
+    ConfigureResponseEventOptions,
+    RequestEvent,
+    ResponseEvent
+} from 'testcafe-hammerhead';
+import { castArray } from 'lodash';
```
Comment (AndreyBelym) [inline src/api/request-hooks/hook.ts:8]: Could you please separate multiline imports with an empty line?

```suggestion

import { castArray } from 'lodash';
```

File: src/api/request-hooks/hook.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import { RequestHookNotImplementedMethodError } from '../../errors/test-run';
+import { RequestFilterRuleInit } from './interfaces';
+import WarningLog from '../../notifications/warning-log';
+
+export default abstract class RequestHook {
```
Comment (AndreyBelym) [inline src/api/request-hooks/hook.ts:13]:
```suggestion

export default abstract class RequestHook {
```

File: src/api/request-hooks/interfaces.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+}
+
+export interface RequestHookLogOptions {
+    logRequestHeaders: boolean;
+    logRequestBody: boolean;
+    stringifyRequestBody: boolean;
+    logResponseHeaders: boolean;
+    logResponseBody: boolean;
+    stringifyResponseBody: boolean;
+}
```
Comment (AndreyBelym) [inline src/api/request-hooks/interfaces.ts:19]:
```suggestion
export interface RequestHookLogOptions {
    logRequestHeaders: boolean;
    logRequestBody: boolean;
    stringifyRequestBody: boolean;
    logResponseHeaders: boolean;
    logResponseBody: boolean;
    stringifyResponseBody: boolean;
}

export type RequestHookLogOptionsInit = Partial<RequestHookLogOptions>;
```

**Comment by AndreyBelym** [comment]
We can utilize ESLint to enforce newlines after import: https://github.com/DevExpress/testcafe/issues/5391#issuecomment-689562436

**Comment by miherlosev** [comment]
> We can utilize ESLint to enforce newlines after import: #5391 (comment)

I will do this in separate PR.

**Comment by miherlosev** [comment]
@testcafe-build-bot retest

**Comment by miherlosev** [comment]
FPR

File: src/api/request-hooks/request-logger.ts [hunk 4]
```diff
@@ -1,12 +1,22 @@
-import { ConfigureResponseEventOptions } from 'testcafe-hammerhead';
+import {
+    ConfigureResponseEventOptions,
+    RequestEvent,
+    ResponseEvent
+} from 'testcafe-hammerhead';
```
Comment (AndreyBelym) [inline src/api/request-hooks/request-logger.ts:5]:
```suggestion

import RequestHook from './hook';
```

File: src/api/request-hooks/request-logger.ts [hunk 5]
```diff
@@ -1,12 +1,22 @@
...
 import RequestHook from './hook';
 import parseUserAgent from '../../utils/parse-user-agent';
 import testRunTracker from '../test-run-tracker';
 import ReExecutablePromise from '../../utils/re-executable-promise';
 import { APIError } from '../../errors/runtime';
 import { RUNTIME_ERRORS } from '../../errors/types';
-
-const DEFAULT_OPTIONS = {
+import {
```
Comment (AndreyBelym) [inline src/api/request-hooks/request-logger.ts:12]:
```suggestion

import {
```

File: src/api/request-hooks/request-logger.ts [hunk 6]
```diff
@@ -1,12 +1,22 @@
...
+    RequestFilterRuleInit,
+    RequestHookLogOptionsInit,
+    RequestHookLogOptions
+} from './interfaces';
+import { Dictionary } from '../../configuration/interfaces';
```
Comment (AndreyBelym) [inline src/api/request-hooks/request-logger.ts:17]:
```suggestion

import { Dictionary } from '../../configuration/interfaces';
```

File: src/api/request-hooks/request-logger.ts [hunk 7]
```diff
@@ -1,12 +1,22 @@
...
+
+const DEFAULT_OPTIONS: RequestHookLogOptions = {
```
Comment (AndreyBelym) [inline src/api/request-hooks/request-logger.ts:19]:
```suggestion

const DEFAULT_OPTIONS: RequestHookLogOptions = {
```

File: src/api/request-hooks/request-mock.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
+import RequestHook from './hook';
+import {
```
Comment (AndreyBelym) [inline src/api/request-hooks/request-mock.ts:2]:
```suggestion

import {
```

File: src/api/request-hooks/request-mock.ts [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    ResponseMock,
+    RequestEvent,
+    ResponseEvent,
+    RequestFilterRule,
+    SAME_ORIGIN_CHECK_FAILED_STATUS_CODE
+} from 'testcafe-hammerhead';
+import { APIError } from '../../errors/runtime';
```
Comment (AndreyBelym) [inline src/api/request-hooks/request-mock.ts:9]:
```suggestion

import { APIError } from '../../errors/runtime';
```

File: src/api/request-hooks/request-mock.ts [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+import { RUNTIME_ERRORS } from '../../errors/types';
+import WARNING_MESSAGE from '../../notifications/warning-message';
+import WarningLog from '../../notifications/warning-log';
+import { RequestFilterRuleInit } from './interfaces';
+
+class RequestMock extends RequestHook {
```
Comment (AndreyBelym) [inline src/api/request-hooks/request-mock.ts:15]:
```suggestion

class RequestMock extends RequestHook {
```

File: src/api/request-hooks/request-logger.ts [hunk 11]
```diff
@@ -1,12 +1,26 @@
-import { ConfigureResponseEventOptions } from 'testcafe-hammerhead';
+import {
+    ConfigureResponseEventOptions,
+    RequestEvent,
+    ResponseEvent
+
```
Comment (AndreyBelym) [inline src/api/request-hooks/request-logger.ts:5]:
```suggestion
} from 'testcafe-hammerhead';

```

**Pull request merged**